src/methods.jl

```diff
@@ -260,16 +271,44 @@ function _rand(rng, p_cumulative, R)
     return index
 end
 
-Random.eltype(::Type{<:UnivariateFinite{<:Any,V}}) where V = V
+
+# # RAND
+
+Random.eltype(::Type{<:UnivariateFinite{S,V,R}}) where {S,V,R} =
+    CategoricalArrays.CategoricalValue{V,R}
 
 # The Sampler hook into Random's API is discussed in the Julia documentation, in the
 # Standard Library section on Random.
+
+
+## Single samples
+
+Random.Sampler(::AbstractRNG, d::UnivariateFinite, ::Val{1}) = Random.SamplerTrivial(d)
+
+function Base.rand(
+    rng::AbstractRNG,
+    sampler::Random.SamplerTrivial{<:UnivariateFinite{<:Any,<:Any,<:Any,P}},
+    ) where P
+
+    d = sampler[]
+    u = rand(rng)
+
+    total = zero(P)
+    for (i, prob) in enumerate(values(d.prob_given_ref))
+        total += prob
+        u <= total && return fast_support(d)[i]
+    end
+end
```

Comment:
```suggestion
function Base.rand(
    rng::AbstractRNG,
    sampler::Random.SamplerTrivial{<:UnivariateFinite{<:Any,<:Any,<:Any,P}},
    ) where P

    d = sampler[]
    u = rand(rng)

    total = zero(P)
    prob_given_ref = d.prob_given_ref
    # For type stability we assign the first index as the default ref
    # This isn't a problem since we know that `rand` is always defined 
    # as UnivariateFinite objects have non-negative probabilities,
    # summing up to a non-negative value.
    rng_key = first(keys(prob_given_ref)) 
    for (ref, prob) in prob_given_ref
        total += prob
        u <= total && begin
            rng_key = ref
            break
        end
    end
    return d.decoder(rng_key)
end
```

The slight slowdown you are currently experiencing in this PR is due to unnecessary allocations and type instability in the for loop. The code suggestion above should fix this.
```julia
julia> using CategoricalDistributions

julia> using BenchmarkTools

julia> categ = Distributions.Categorical([0.5, 0.4, 0.1])
Categorical{Float64, Vector{Float64}}(support=Base.OneTo(3), p=[0.5, 0.4, 0.1])

julia> d = UnivariateFinite(["maybe", "no", "yes"], [0.5, 0.4, 0.1]; pool=missing)
UnivariateFinite{Multiclass{3}}(maybe=>0.5, no=>0.4, yes=>0.1)

julia> @btime rand($categ) 
  12.012 ns (0 allocations: 0 bytes)
2

julia> @btime rand($d) # formerly 42 ns
  17.535 ns (0 allocations: 0 bytes)
CategoricalArrays.CategoricalValue{String, UInt8} "maybe"

julia> versioninfo()
Julia Version 1.9.3
Commit bed2cd540a (2023-08-24 14:43 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, tigerlake)
  Threads: 1 on 8 virtual cores

julia>